[WPE][GTK] Add damage visualizer

Scony opened the pull request:
#### 5280aba84b7db28f3a2d689c0f9c83a6f7a9c85d
<pre>
[WPE][GTK] Add damage visualizer
<a href="https://bugs.webkit.org/show_bug.cgi?id=286907">https://bugs.webkit.org/show_bug.cgi?id=286907</a>

Reviewed by Carlos Garcia Campos.

This change adds a damage visualizer that can be controlled via environment variable.

The visualizer visualizes the damage rectangles to visually highlight what changed
on the screen during the frame renedering.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp: Added.
(WebCore::TextureMapperDamageVisualizer::TextureMapperDamageVisualizer):
(WebCore::TextureMapperDamageVisualizer::updateDamageAndDisplay):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.h: Added.
(WebCore::TextureMapperDamageVisualizer::isActive const):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:

Canonical link: <a href="https://commits.webkit.org/289785@main">https://commits.webkit.org/289785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb6c08743135cddf23a734e2bf3a67e8a4a6e2f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88047 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7563 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92995 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67966 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25696 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6084 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79656 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48329 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5858 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34066 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37904 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94843 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15215 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76819 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15470 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75512 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76056 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18695 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20452 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18855 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8228 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15233 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20534 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14975 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18420 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16757 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+#include "config.h"
+
+#if ENABLE(DAMAGE_TRACKING)
+#include "TextureMapperDamageVisualizer.h"
```
This should be right after the config include. I doesn't need to be guarded, since the header also has the ifdef.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+            m_margin = *value - 1;
+        }
+    }
+}
```
I know this is how the fps counter work, but I think it would be better to create the class on the heap only when needed. Adding a create() static function where we read the env var/setting and return nullptr if not needed.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+    }
+}
+
+void TextureMapperDamageVisualizer::updateDamageAndDisplay(TextureMapper& textureMapper, const Damage& damage)
```
The FPS is called update and display because the internal counter is updated, but here we don't keep the damage, I would just call this paintDamage or something similar.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+        return;
+
+    const FloatPoint offset(m_margin, m_margin);
+    const FloatSize extraSize(m_margin * 2, m_margin * 2);
```
I think you could use FloatBoxExtent, it could be created by the create() from the env var and passed to the constructor to keep it as a member.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+    for (const auto& rect : damage.rects()) {
+        FloatRect rectToDraw(rect);
+        rectToDraw.moveBy(-offset);
+        rectToDraw.expand(extraSize);
```
Then you can use it here directly, there's a expand() that receives a FloatBoxExtent

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
```diff
@@ -43,6 +43,10 @@
 
+#if ENABLE(DAMAGE_TRACKING)
+#include <WebCore/TextureMapperDamageVisualizer.h>
+#endif
```
Don't include it conditionally, since the header is already guarded.

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+    , m_margin(0)
+{
+    if (const auto* showDmageEnvvar = getenv("WEBKIT_SHOW_DAMAGE")) {
+        const auto value = parseInteger<unsigned>(StringView::fromLatin1(showDmageEnvvar));
```
```suggestion
    if (const auto* showDamageEnvvar = getenv("WEBKIT_SHOW_DAMAGE")) {
        const auto value = parseInteger<unsigned>(StringView::fromLatin1(showDamageEnvvar));
```

(even if the code is moved elsewhere)

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+
+    const FloatPoint offset(m_margin, m_margin);
+    const FloatSize extraSize(m_margin * 2, m_margin * 2);
+    const SRGBA<uint8_t> color(255, 0, 0, 200);
```
`red` is already defined in `WebCore::color::red`, so you could...

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,71 @@
+        FloatRect rectToDraw(rect);
+        rectToDraw.moveBy(-offset);
+        rectToDraw.expand(extraSize);
+        textureMapper.drawSolidColor(rectToDraw, { }, color, true);
```
...possibly do

```suggestion
        textureMapper.drawSolidColor(rectToDraw, { }, WebCore::Color::red.colorWithAlphaByte(200), true);
```

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp:
```diff
@@ -0,0 +1,61 @@
+    if (const auto* showDamageEnvvar = getenv("WEBKIT_SHOW_DAMAGE")) {
+        const auto value = parseInteger<unsigned>(StringView::fromLatin1(showDamageEnvvar));
+        if (value && *value > 0)
+            return std::unique_ptr<TextureMapperDamageVisualizer>(new TextureMapperDamageVisualizer(*value - 1));
```
```suggestion
            return WTF::makeUnique<TextureMapperDamageVisualizer>(*value - 1);
```

Scony replied:
Not possible as `WTF::makeUnique` is not a `friend` of `TextureMapperDamageVisualizer`. I've checked few places and it looks like avoiding befriending is preferred in WebKit

carlosgcampos replied:
Just make the constructor public

Scony replied:
> Just make the constructor public

If we're okay we that then I'll do it.

carlosgcampos replied:
Yes, it's what we always do to be able to use makeUnique

spenap replied:
There's an idiom that addresses this problem: https://playfulprogramming.com/posts/a-forgotten-idiom-revisited-pass-key

It requires a `tag` (what the blog post calls the _pass-key_. The tag is kept private / friend-only, so even if the constructor remains public, clients can't use it and only the `create` method can

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.h:
```diff
@@ -0,0 +1,55 @@
+    void paintDamage(TextureMapper&, const Damage&) const;
+
+private:
+
```
```suggestion
```

Scony committed:
[WPE][GTK] Add damage visualizer
https://bugs.webkit.org/show_bug.cgi?id=286907

Reviewed by Carlos Garcia Campos.

This change adds a damage visualizer that can be controlled via environment variable.

The visualizer visualizes the damage rectangles to visually highlight what changed
on the screen during the frame renedering.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.cpp: Added.
(WebCore::TextureMapperDamageVisualizer::TextureMapperDamageVisualizer):
(WebCore::TextureMapperDamageVisualizer::updateDamageAndDisplay):
* Source/WebCore/platform/graphics/texmap/TextureMapperDamageVisualizer.h: Added.
(WebCore::TextureMapperDamageVisualizer::isActive const):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::paintToCurrentGLContext):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:

Canonical link: https://commits.webkit.org/289785@main
